src/lib/socks/headers.ts

```diff
@@ -202,12 +215,75 @@ export function interpretAuthResponse(buffer:ArrayBuffer) : Auth {
 
   // Only SOCKS Version 5 is supported.
   var socksVersion = byteArray[0];
-  if (socksVersion != VERSION5) {
+  if (socksVersion != Version.VERSION5) {
     throw new Error('unsupported SOCKS version: ' + socksVersion);
   }
   return byteArray[1];
 }
 
+// In between Steps 2 and 3, additional subnegotiation messages may need to
+// be exchanged depending on which authentication method was negotiated.
+//
+// Client to Server (USERPASS Subnegotiation)
+// https://tools.ietf.org/html/rfc1929
+//
+// Given a username and password combination for userpass auth, compose and
+// interpret the auth request.
+//
+//            +----+------+----------+------+----------+
+//            |VER | ULEN |  UNAME   | PLEN |  PASSWD  |
+//            +----+------+----------+------+----------+
+//            | 1  |  1   | 1 to 255 |  1   | 1 to 255 |
+//            +----+------+----------+------+----------+
+export function composeUserPassRequest(creds:UserPass) : ArrayBuffer {
+  var ulen :number = 2 * creds.username.length; // 2 bytes per char
```

Comment:
No need for `:number` since the compiler can easily infer the type. Also, if something can be `const` it generally should.